sanity-io/gatsby-source-sanity
# Gatsby: Content sync changes

This adds functionality for the source plugin to create node manifests. These manifests are used in tandem with a new feature we are rolling out called Content Sync. This will make it so we can link content in a CMS that a user wants to preview to the exact page and build that the user wants to preview and route the user to that content as soon as it's ready. Previously if the page was not created yet it would 404, leading to a confusing experience for the CMS user.

These source plugin changes will be accompanied by changes to the official Gatsby starters to improve Gatsby’s Preview service on Sanity. The user will be redirected to a loading page on Gatsby Cloud when they press “Open Preview”. 
![Screen Shot 2021-11-12 at 1 28 26 PM](https://user-images.githubusercontent.com/49894658/141846361-d4623aa4-9ff2-42d7-816e-c7da9c5e42a6.png)

The loading page waits for the associated build to complete and then redirects the user to the correct page once the updated content is available. The loading page also handles build errors and will let the user know if the build they are waiting for fails.


The updated plugin code and test website are located here for reference: https://github.com/gatsby-inc/sanity-content-sync. The source plugin changes are in the plugins directory of this Gatsby site. 

To test out this functionality: 
1. Copy the code from the source plugins folder @ https://github.com/gatsby-inc/sanity-content-sync/tree/main/web to your Gatsby test site. Upgrade to `gatsby@next`
2. Host the test site on Gatsby Cloud and ensure CMS preview is enabled.  
3. Use a Sanity studio project with webhooks enabled for CMS preview for testing. 
4. Pending updates to the official Gatsby starter to include the Preview extension outlined above, you can now either test by: 
        a. Incorporating the document action component defined in `gatsbyPreviewAction.js` @ 
https://github.com/gatsby-inc/sanity-content-sync/tree/main/studio into your studio site and setting the `contentSyncURL` as an environmental variable. https://gatsbyjs.com/content-sync/#{site_id}
        b. OR navigate to the content sync URL manually after making content changes in Sanity Studio. 
  https://gatsbyjs.com/content-sync/#{site_id}/gatsby-source-sanity/#{manifestid} 
The manifest Id can be found in the logs of the source plugin code when a content update is made. For example: 
<img width="250" alt="Screen Shot 2021-11-15 at 12 00 10 PM" src="https://user-images.githubusercontent.com/49894658/141846013-5bfd9454-4128-4dd4-8744-b325a3310be0.png">

Note: These changes are inclusive of the recent PINC build updates.
